Add an FAQ entry for limiting NumPy threads

StefansM opened the pull request:
This pull request addresses issue #1397 by adding an entry to the FAQ giving some workarounds to the problem.

I won't be offended if you ask me to rewrite it: I tend to be a bit too verbose when it comes to this sort of thing.

---

I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ AND the _BSD 3-Clause License_.

I am happy to be thanked by name in the ``NEWS.rst`` and
``CONTRIB.rst`` files.

I have read the ``CONTRIBUTING.rst`` file and understand that AppVeyor and
TravisCI will be used to confirm the Biopython unit tests and ``flake8`` style
checks pass with these changes.

peterjc commented on Doc/Tutorial/chapter_introduction.tex:
```diff
@@ -312,6 +312,24 @@ \section{Frequently Asked Questions (FAQ)}
+  \begin{verbatim}
+  import os
+  try:
+    os.environ["OMP_NUM_THREADS"] = "1"
```
Four space indentation please

peterjc commented on Doc/Tutorial/chapter_introduction.tex:
```diff
@@ -312,6 +312,24 @@ \section{Frequently Asked Questions (FAQ)}
   We deprecated the \verb|Bio.Fasta| module in Biopython 1.51 (August 2009) and removed it in Biopython 1.55 (August 2010). There is a brief example showing how to convert old code to use \verb|Bio.SeqIO| instead in the \href{https://github.com/biopython/biopython/blob/master/DEPRECATED.rst}{DEPRECATED.rst} file.
 
+  \item \emph{Why does Biopython start so many threads?} \\
+  Some parts of Biopython make use of \href{http://www.numpy.org}{NumPy}, which in turn makes use of a system-dependent BLAS (Basic Linear Algebra Subprograms) library. Some BLAS libraries automatically spawn one thread for each available CPU core. On systems where this is the case, simply importing NumPy (or any Biopython module relying on NumPy) will spawn several threads. The number of threads ...
```
Should we say *"... which in turn may make use of ..."* here (ie *may*)? After all, NumPy might be built with ATLAS instead of BLAS?

peterjc commented:
@StefansM Other than those very minor tweaks, this looks good. I've double checked the LaTeX rendering locally (since the verbatim section can be tricky given all the special characters), and that's perfect.

@mdehoon you do a lot more with NumPy than me - any comments on this?

StefansM commented:
@peterjc I've fixed the indentation of the python sample. I should know better than to trust autoindent for code embedded in LaTeX documents.

I also reworded the text slightly following your suggestion. The Intel MKL isn't just a BLAS library, so I changed that to "mathematics library".

peterjc commented:
Looks good. If there are no objections or further suggestions, I'll merge this later this week.

peterjc commented:
Apparently it can be ``$OPENBLAS_NUM_THREADS`` which needs setting, see https://mail.python.org/pipermail/numpy-discussion/2017-September/077224.html

StefansM commented:
It looks like it can be `OPENBLAS_NUM_THREADS`, `GOTO_NUM_THREADS` or `OMP_NUM_THREADS`, depending on how OpenBLAS was linked: https://github.com/xianyi/OpenBLAS#set-the-number-of-threads-with-environment-variables

I think these are the most common libraries that NumPy is linked with:

* **Netlib LAPACK/BLAS**: Single-threaded only, nothing to set.
* **ATLAS**: [Set at compile time](http://math-atlas.sourceforge.net/faq.html#tnum), nothing that can be done.
* **BLIS**: Set [`BLIS_NUM_THREADS`](https://github.com/flame/blis/wiki/Multithreading#the-automatic-way)
* **Intel MKL**: Set [`MKL_NUM_THREADS`](https://software.intel.com/en-us/articles/recommended-settings-for-calling-intel-mkl-routines-from-multi-threaded-applications)
* [**OpenBLAS**](https://github.com/xianyi/OpenBLAS#set-the-number-of-threads-with-environment-variables):
  + Compiled with OpenMP: Set `OMP_NUM_THREADS`
  + Otherwise, set `OPENBLAS_NUM_THREADS` or `GOTO_NUM_THREADS`.
  + Practically speaking, it's probably best to set both `OPENBLAS_NUM_THREADS` and `OMP_NUM_THREADS`.

I've not actually tested whether this problem occurs with these other libraries, but it seems likely. I'm happy to add these to the FAQ entry if you think it wouldn't bulk it out too much.

peterjc commented:
Let's give it a day or two in case that thread on the NumPy discussion list has a more authoritative answer (e.g. in some cases is it function call time rather than import time which matters?)

peterjc commented:
We never did merge this - and surprisingly I don't recall any queries in the years since about threads via numpy. I'm going to just close this, sorry, rebasing is non-trivial.

Note the tutorial has since moved from LaTeX to reStructuredText and the FAQ is now here:
https://github.com/biopython/biopython/blob/biopython-184/Doc/Tutorial/chapter_introduction.rst
